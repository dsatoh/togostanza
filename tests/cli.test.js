import os from 'os';
import path from 'path';
import { spawnSync } from 'child_process';

import { copySync, mkdtempSync, pathExistsSync, readFileSync, removeSync } from 'fs-extra';

test('--help', () => {
  const {stdout} = togostanza('--help');

  expect(stdout).toContain('Usage: togostanza [options] [command]');
});

describe('init', () => {
  test.each([
    {packageManager: 'npm'},
    {packageManager: 'yarn'}
  ])('non-interactive (%p)', ({packageManager}) => {
    withinTmpdir(() => {
      const {status, output} = togostanza(
        'init', 'some-repo',
        '--license',         'MIT',
        '--package-manager', packageManager,
        '--owner',           'ursm',
        '--repo',            'some-repo',
        '--skip-install',
        '--skip-git'
      );

      expect(status).toBe(0);
      expect(output).toMatchSnapshot();

      expect(readFileSync('some-repo/README.md', 'utf8')).toMatchSnapshot();
      expect(readFileSync('some-repo/package.json', 'utf8')).toMatchSnapshot();
      expect(readFileSync('some-repo/.github/workflows/publish.yml', 'utf8')).toMatchSnapshot();
    });
  });
});

describe('generate stanza', () => {
  test('non-interactive', () => {
    withinTmpdir(() => {
      const {status, output} = togostanza(
        'generate', 'stanza', 'hello',
        '--label',      'LABEL',
        '--definition', 'DEFINITION',
        '--type',       'TYPE',
        '--context',    'CONTEXT',
        '--display',    'DISPLAY',
        '--provider',   'PROVIDER',
        '--license',    'LICENSE',
        '--author',     'AUTHOR',
        '--address',    'ADDRESS'
      );

      expect(status).toBe(0);
      expect(output).toMatchSnapshot();

      expect(readFileSync('stanzas/hello/metadata.json', 'utf8')).toMatchSnapshot();
      expect(readFileSync('stanzas/hello/index.js', 'utf8')).toMatchSnapshot();
      expect(readFileSync('stanzas/hello/templates/stanza.html.hbs', 'utf8')).toMatchSnapshot();
    });
  });
});

describe('upgrade', () => {
  test('move stanza into stanzas/', () => {
    withinTmpdir((here) => {
      copySync(path.resolve(__dirname, 'fixtures/upgrade/move-stanza-into-stanzas'), here);

      expect(pathExistsSync('hello/metadata.json')).toBe(true);

      const {status} = togostanza('upgrade');

      expect(status).toBe(0);

      expect(pathExistsSync('hello/metadata.json')).toBe(false);
      expect(pathExistsSync('stanzas/hello/metadata.json')).toBe(true);
    });
  });
});

function togostanza(...args) {
  const bin = path.resolve(__dirname, '../bin/togostanza.mjs');

  // --no-warnings is necessary to avoid including ESM warnings in the snapshot, which change with each test run
  // (not required in Node.js 14)
  return spawnSync(process.argv0, ['--no-warnings', bin, ...args], {encoding: 'utf8', timeout: 3000});
}

function withinTmpdir(cb) {
  const tmp = mkdtempSync(path.join(os.tmpdir(), 'togostanza-'));

  try {
    const cwd = process.cwd();

    process.chdir(tmp);

    try {
      cb(tmp);
    } finally {
      process.chdir(cwd);
    }
  } finally {
    removeSync(tmp);
  }
}
