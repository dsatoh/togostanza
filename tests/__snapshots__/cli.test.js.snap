// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build simple 1`] = `
Array [
  null,
  "",
  "   create stanzas/hello/metadata.json
   create stanzas/hello/index.js
   create stanzas/hello/templates/stanza.html.hbs
",
]
`;

exports[`build simple 2`] = `
Array [
  "-togostanza/",
  "-togostanza/Layout-HASH.js",
  "-togostanza/Layout-HASH.js.map",
  "-togostanza/help-app.js",
  "-togostanza/help-app.js.map",
  "-togostanza/index-app.js",
  "-togostanza/index-app.js.map",
  "hello.html",
  "hello.js",
  "hello.js.map",
  "index.html",
]
`;

exports[`generate stanza non-interactive 1`] = `
Array [
  null,
  "",
  "   create stanzas/hello/metadata.json
   create stanzas/hello/index.js
   create stanzas/hello/templates/stanza.html.hbs
",
]
`;

exports[`generate stanza non-interactive 2`] = `
Object {
  "stanzas": Object {
    "hello": Object {
      "index.js": "export default function hello(stanza, params) {
  const sayTo = params['say-to'] || 'world';

  stanza.render({
    template: 'stanza.html',
    parameters: {
      greeting: \`Hello, \${sayTo}!\`
    }
  });
}
",
      "lib": Object {},
      "metadata.json": "{
  \\"@context\\": {
    \\"stanza\\": \\"http://togostanza.org/resource/stanza#\\"
  },
  \\"@id\\": \\"hello\\",
  \\"stanza:label\\": \\"LABEL\\",
  \\"stanza:definition\\": \\"DEFINITION\\",
  \\"stanza:parameter\\": [
    {
      \\"stanza:key\\": \\"say-to\\",
      \\"stanza:example\\": \\"world\\",
      \\"stanza:description\\": \\"who to say hello to\\",
      \\"stanza:required\\": false
    }
  ],
  \\"stanza:about-link-placement\\": \\"bottom-right\\",
  \\"stanza:style\\": [],
  \\"stanza:usage\\": \\"<togostanza-hello></togostanza-hello>\\",
  \\"stanza:type\\": \\"TYPE\\",
  \\"stanza:context\\": \\"CONTEXT\\",
  \\"stanza:display\\": \\"DISPLAY\\",
  \\"stanza:provider\\": \\"PROVIDER\\",
  \\"stanza:license\\": \\"LICENSE\\",
  \\"stanza:author\\": \\"AUTHOR\\",
  \\"stanza:address\\": \\"ADDRESS\\",
  \\"stanza:contributor\\": [],
  \\"stanza:created\\": \\"2020-10-05\\",
  \\"stanza:updated\\": \\"2020-10-05\\"
}
",
      "templates": Object {
        "stanza.html.hbs": "<p>{{greeting}}</p>
",
      },
    },
  },
}
`;

exports[`init non-interactive ({"packageManager": "npm"}) 1`] = `
Array [
  null,
  "",
  "   create package.json
   create README.md
   create .gitignore
   create .github/workflows/publish.yml
Skipping install command: npm install --cache-min 86400


I'm all done. Just run npm install to install the required dependencies.



Getting Started
---------------

Create a new stanza:

  $ cd some-repo
  $ npx togostanza generate stanza

Serve the repository locally:

  $ cd some-repo
  $ npx togostanza serve

Build stanzas for deployment:

  $ cd some-repo
  $ npx togostanza build

",
]
`;

exports[`init non-interactive ({"packageManager": "npm"}) 2`] = `
Object {
  "some-repo": Object {
    ".github": Object {
      "workflows": Object {
        "publish.yml": "name: Publish GitHub Pages

on:
  push:
    branches:
      - master

jobs:
  deploy:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v2
        with:
          submodules: true
          fetch-depth: 0

      - uses: actions/setup-node@v1
        with:
          node-version: '12'

      - run: npm ci
      - run: npx togostanza build

      - uses: peaceiris/actions-gh-pages@v3
        with:
          github_token: \${{ secrets.GITHUB_TOKEN }}
          publish_dir: ./dist
",
      },
    },
    ".gitignore": "/dist/
/node_modules/
",
    "README.md": "# some-repo
",
    "lib": Object {},
    "package.json": "{
  \\"name\\": \\"some-repo\\",
  \\"version\\": \\"0.0.1\\",
  \\"license\\": \\"MIT\\",
  \\"repository\\": \\"https://github.com/ursm/some-repo.git\\",
  \\"dependencies\\": {
    \\"togostanza\\": \\"github:togostanza/togostanza-js\\"
  },
  \\"engines\\": {
    \\"node\\": \\">=12\\"
  },
  \\"private\\": true
}
",
  },
}
`;

exports[`init non-interactive ({"packageManager": "yarn"}) 1`] = `
Array [
  null,
  "",
  "   create package.json
   create README.md
   create .gitignore
   create .github/workflows/publish.yml
Skipping install command: yarn install


I'm all done. Just run yarn install to install the required dependencies.



Getting Started
---------------

Create a new stanza:

  $ cd some-repo
  $ yarn run togostanza generate stanza

Serve the repository locally:

  $ cd some-repo
  $ yarn run togostanza serve

Build stanzas for deployment:

  $ cd some-repo
  $ yarn run togostanza build

",
]
`;

exports[`init non-interactive ({"packageManager": "yarn"}) 2`] = `
Object {
  "some-repo": Object {
    ".github": Object {
      "workflows": Object {
        "publish.yml": "name: Publish GitHub Pages

on:
  push:
    branches:
      - master

jobs:
  deploy:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v2
        with:
          submodules: true
          fetch-depth: 0

      - uses: actions/setup-node@v1
        with:
          node-version: '12'

      - run: yarn install --frozen-lockfile
      - run: yarn run togostanza build

      - uses: peaceiris/actions-gh-pages@v3
        with:
          github_token: \${{ secrets.GITHUB_TOKEN }}
          publish_dir: ./dist
",
      },
    },
    ".gitignore": "/dist/
/node_modules/
",
    "README.md": "# some-repo
",
    "lib": Object {},
    "package.json": "{
  \\"name\\": \\"some-repo\\",
  \\"version\\": \\"0.0.1\\",
  \\"license\\": \\"MIT\\",
  \\"repository\\": \\"https://github.com/ursm/some-repo.git\\",
  \\"dependencies\\": {
    \\"togostanza\\": \\"github:togostanza/togostanza-js\\"
  },
  \\"engines\\": {
    \\"node\\": \\">=12\\"
  },
  \\"private\\": true
}
",
  },
}
`;

exports[`upgrade move stanza into stanzas/ 1`] = `
"move  hello -> stanzas/hello

No files selected, nothing to do. 
All done. 
Results: 
0 errors
0 unmodified
0 skipped
0 ok

"
`;

exports[`upgrade move stanza into stanzas/ 2`] = `""`;

exports[`upgrade move stanza into stanzas/ 3`] = `
Object {
  "stanzas": Object {
    "hello": Object {
      "metadata.json": "{}",
    },
  },
}
`;
